pkg/controller/job/backoff_utils.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package job
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/utils/clock"
+)
+
+type backoffRecord struct {
```

Comment:
Have you looked into hiding the `backoffRecord` as an implementation details of `backoffStore`? 
Such that a consumer of this would only ever invoke `Add` or `Remove` and the underlying mechanism would be smart enough to figure out if `Add` is updating existing resource with new data, or creating entirely new. I think this could potentially improve the API surface of this structure. Just a thought :wink:

Reply:
Thank you @soltysh! Will have a look..